doc/jacoco.md

```diff
@@ -0,0 +1,164 @@
+
+First, we still have to use `prepare-agent` goal as we did in basic example in all modules, the best way to achieve this is to configure the execution of this goal in the parent `pom.xml`. Then we need to add execution to the module containing integration tests to generate the report across modules
+
+```
+<build>
+  <plugins>
+    <plugin>
+      <groupId>org.jacoco</groupId>
+      <artifactId>jacoco-maven-plugin</artifactId>
+      <executions>
+        <execution>
+          <id>report</id>
+          <goals>
+            <goal>report-aggregate</goal>
+          </goals>
+          <phase>verify</phase>
+        </execution>
+      </executions>
+    </plugin>
+  </plugins>
+</build>
+```
+
+We snippet binds the goal to the `verify` phase. Now when we execute
+
+```
+mvn clean verify
+```
+
+we should find the aggregated report in `target/site/jacoco-aggregate/jacoco.xml` of the module containing the integration tests.
+
+However, the JaCoCo plugin imports coverage reports module by module, so we need to import the same report multiple times (once for every module) to have coverage for all modules. To achieve this we set the property `sonar.coverage.jacoco.xmlReportPaths` in every module 
+
+```
+<properties>
+  <sonar.coverage.jacoco.xmlReportPaths>${project.basedir}/../integration_tests/target/site/jacoco-aggregate/jacoco.xml</sonar.coverage.jacoco.xmlReportPaths>
+</properties>
+```
+
+We can factor out the path to the report in the property of the parent `pom.xml` to avoid repetition.
+
+### Troubleshooting
+
+To investigate issues with the import of coverage information you can run Maven with the debug flag, `-X`:
+
+```
+mvn -X clean verify sonar:sonar 
+```
+
+In the logs, you will find the execution of different sensors for each module of the project. Typically you will have a log similar to the following one when XML report is processed.
+
+```
+[INFO] 16:58:05.074 Sensor JaCoCo XML Report Importer [jacoco]
+[DEBUG] 16:58:05.082 Reading report 'C:\projects\sonar-scanning-examples\maven-multimodule\tests\target\site\jacoco-aggregate\jacoco.xml'
+[INFO] 16:58:05.093 Sensor JaCoCo XML Report Importer [jacoco] (done) | time=19ms
+```
+
+
+
+## Gradle
+
+Gradle includes the [JaCoCo plugin](https://docs.gradle.org/current/userguide/jacoco_plugin.html) in its default distribution. To apply the plugin to a project, you need to declare it in your `build.gradle` file together with the [SonarScanner for Gradle](https://docs.sonarqube.org/latest/analysis/scan/sonarscanner-for-gradle/).
+
+```
+plugins {
+    id "jacoco"
+    id "org.sonarqube" version "2.7.1"
+}
+```
+
+The plugin provides the `jacocoTestReport` task, which needs to be configured to produce an XML report.
+
+```
+jacocoTestReport {
+    reports {
+        xml.enabled true
+        csv.enabled false
```

Comment:
Why?

Reply:
I assume you asked why I disabled other formats - no particular reason. I removed disabling other formats

Reply:
Sorry I wasn't clear. No, I'm asking why we would explicitly enable csv generation. It's not human readable like the html report, nor useful for SQ analysis.

Reply:
ah, I think you are reading it wrong, this was enabling `xml` and disabling `csv` and `html` - notice the `true` / `false`  which is setting the property. (I know - groovy :) )

Reply:
:woman_facepalming: